Fix RFLink unit conversions

javicalle opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Recent changes to unit conversions and other refactorings have caused problems in the RFLink integration when the units of measure and display units do not match.
This PR adds the `device_class` attribute for those entities that require unit conversions.

It has also been necessary to define the `unique_id` property so that the behavior of the sensors is adequate.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->
In #69303, had been asked to migrate from `self._device_id` to use `self.unique_id`:
* https://github.com/home-assistant/core/pull/69303#discussion_r881411532

I can do the refactor but I prefer to do it in a separate PR.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
  - #73920 
  - #69303
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

javicalle committed:
Fix unit conversions

javicalle committed:
Fix tests

elupus commented on homeassistant/components/rflink/sensor.py:
```diff
@@ -38,6 +42,15 @@
+    "windspeed": SensorDeviceClass.SPEED,
+    "temperature": SensorDeviceClass.TEMPERATURE,
+    # SensorDeviceClass.VOLUME,
+    # SensorDeviceClass.WEIGHT,
```
Don't add commented out code.

elupus commented on homeassistant/components/rflink/__init__.py:
```diff
@@ -372,6 +372,11 @@ def _handle_event(self, event):
+    @property
+    def unique_id(self):
+        """Return the unique ID of this entity."""
+        return self._device_id
```
This looks a bit strange. Cant there be multiple sensors per device?

javicalle replied:
No.
TL;DR;
The RFLink library split every device report (with multiple sensors measurements) in diferents sensor reports (one for measurement).
That `_device_id` was an unlucky name for the attribute.
In previous refactorings I tried to change all references to `entity_id` myself, but the PR was ultimately not accepted.
So, that `device_id` refers to the entity id and not the device (that yes, can have multiple sensors)

elupus replied:
This should be set in init, and use the self._attr_unique_id

mbo18 commented:
Can you also add `state_class: measurement` ?

javicalle commented:
> Can you also add `state_class: measurement` ?

Can you give me some sensor example for this?

javicalle commented:
RFLink sensors and units are defined here:
* https://github.com/aequitas/python-rflink/blob/master/rflink/parser.py

User can override most of the units.

javicalle committed:
Address requests changes

javicalle committed:
Address requests changes

javicalle commented:
Added a few more `device_clas` that requires unit conversion. Not sure if they are all the supported from RFLink.
Added the `state_class` attribute just for these entities. I think the correct implementation is to use `SensorEntityDescription` but in my opinion it is beyond this PR.

javicalle committed:
Address requests changes

javicalle committed:
Address requests changes

elupus reviewed (changes requested):
Actually SensorEntityDescription is not beyond the scope here. Adding lookups of measurement and device classes is the very much the point of those descriptions. So it should be done right away.

javicalle commented:
> Actually SensorEntityDescription is not beyond the scope here. Adding lookups of measurement and device classes is the very much the point of those descriptions. So it should be done right away.

IMHO that PR is intended to address the issues that some users have related to temperature conversions.

I will try to implement the `SensorEntityDescription` approach, but I'm not confortable implementing it just for the affected entities (temperature entities) and implement it for all the entities types will require a deep analisys of the RFLink types vs HA types.

elupus commented:
You only need to set the identical things as done in this pull request. That would be a good start

javicalle committed:
migrate to SensorEntityDescription

javicalle committed:
forgotten test

javicalle committed:
forgotten test

javicalle commented:
Implementation migrated to use `SensorEntityDescription` for entities.
Quite simple at the end but I had to understand and implement the new approach.

javicalle commented:
Not sure if I understand the codecov complains. Not seems related to my code.

elupus commented on homeassistant/components/rflink/sensor.py:
```diff
@@ -116,7 +183,9 @@ def __init__(
         """Handle sensor specific args and super init."""
         self._sensor_type = sensor_type
-        self._unit_of_measurement = unit_of_measurement
+        self._attr_native_unit_of_measurement = unit_of_measurement
```
But that in the case of not having an entity description

javicalle replied:
The intention is that user can overwrite the sensor UOM, if not the EntityDescriptor will be taken, if no ED then no UOM.
```python
    @property
    def native_unit_of_measurement(self) -> str | None:
        """Return the unit of measurement of the sensor, if any."""
        if hasattr(self, "_attr_native_unit_of_measurement"):
            return self._attr_native_unit_of_measurement
        if hasattr(self, "entity_description"):
            return self.entity_description.native_unit_of_measurement
        return None
```

Just to be sure, is there (in my implementation) something that may I change?

elupus replied:
There is no fallback to entity description. If the attr exist, that is what will always be used. If you need fallback you need a property to check if the attribute is none. The standard check is if it exists at all.

javicalle replied:
Thanks Joakim, I appreciate your comments.
But I have some doubts if this part is going the way I think.

My intention is to have a 'default' unit of measurement for every `sensor_type`, but the user could always to define the  right one for its device.
For example, a temperature sensor would be 'always' in celsius, but the user must be able to device it in fahrenheit if the device returns this UOM. 

Sorry for asking you about this, but I can't find something similar in the code and that part worries to me.

elupus replied:
Then keep the property for native unit of measurement, if the attribute is none none, return that, if it is none retunr the value from the description.

elupus commented on homeassistant/components/rflink/sensor.py:
```diff
@@ -35,9 +43,68 @@
+        key="average_windspeed",
+        name="Average windspeed",
+        device_class=SensorDeviceClass.SPEED,
+        state_class=SensorStateClass.MEASUREMENT,
```
While we where working a new wind speed device class was added.

javicalle replied:
Changed here and the other ones

elupus commented on homeassistant/components/rflink/sensor.py:
```diff
@@ -35,9 +43,68 @@
+    ),
+    SensorEntityDescription(
+        key="windspeed",
+        name="Windspeed",
```
Wind speed

javicalle replied:
`Average wind speed` too?

elupus replied:
Yes i think so.

javicalle replied:
also changed

elupus reviewed (commented):
Some minor nits.

javicalle committed:
chango to WIND_SPEED

javicalle committed:
manage UOM

javicalle committed:
fix test

elupus commented on homeassistant/components/rflink/sensor.py:
```diff
@@ -164,7 +233,10 @@ async def async_added_to_hass(self) -> None:
+        if self._unit_of_measurement:
+            return self._unit_of_measurement
+        if hasattr(self, "entity_description"):
+            return self.entity_description.native_unit_of_measurement
```
Looks to be missing an explict return None in the failing case.

javicalle replied:
Done.

javicalle committed:
address comments

javicalle commented:
There is something wrong with the implementation.
Luck of the tests.

javicalle committed:
fix implementation

javicalle commented:
Fixed setting always a UOM despite of having a SensorEntityDescription for the entity:
```python
        if ATTR_UNIT_OF_MEASUREMENT not in config:
            if (
                config[CONF_SENSOR_TYPE] not in SENSOR_TYPES_DICT
            ):  # only if not SensorEntityDescription
                config[ATTR_UNIT_OF_MEASUREMENT] = lookup_unit_for_sensor_type(
                    config[CONF_SENSOR_TYPE]
                )
```

It requires to make optional the UOM n `__init__`:
```python
class RflinkSensor(RflinkDevice, SensorEntity):
    """Representation of a Rflink sensor."""

    def __init__(
        self,
        device_id: str,
        sensor_type: str,
        unit_of_measurement: str | None = None,
        initial_event=None,
        **kwargs: Any,
    ) -> None:
```

elupus commented:
Why is lookup_unit_for_sensor_type needed? We should have entity descriptions for all of those?

javicalle commented:
> Why is lookup_unit_for_sensor_type needed?

RFLink defines a UOM for some of the entities:
https://github.com/aequitas/python-rflink/blob/b2dc51041e67b2b2e954d60c0ef04e9986a9b66d/rflink/parser.py#L153-L192

> We should have entity descriptions for all of those?

Yes that would be possible, but again the PR is growing beyond the initial goal (fix the conversion problem presented by the temperature sensors).
In this PR I have tried to focus only on the entities that may present problems with unit conversions, but if it is considered appropriate I can extend it to the rest of the sensors.

elupus commented on homeassistant/components/rflink/sensor.py:
```diff
@@ -73,9 +142,12 @@ def devices_from_config(domain_config):
+            if (
+                config[CONF_SENSOR_TYPE] not in SENSOR_TYPES_DICT
+            ):  # only if not SensorEntityDescription
+                config[ATTR_UNIT_OF_MEASUREMENT] = lookup_unit_for_sensor_type(
```
While this is old code. It mutates the config, which is somewhat vad.
Cant we lookup the unit inside constructor instead if passed unit is none and there is no entity description. Alternatively construct an entity description on the fly using this unit lookup.

javicalle replied:
Done

elupus commented:
> 
> > We should have entity descriptions for all of those?
> 
> Yes that would be possible, but again the PR is growing beyond the initial goal (fix the conversion problem presented by the temperature sensors). In this PR I have tried to focus only on the entities that may present problems with unit conversions, but if it is considered appropriate I can extend it to the rest of the sensors.

Im okey with keeping that type of fallback code for now. Though would be nice to get rid off at some point.

javicalle committed:
prevent from mutate config

javicalle commented:
> Though would be nice to get rid off at some point.

I agree and will be happy to give a try, but in the past a lot of changes has been rejected because the library must be refactorized (and that is something beyond my capabilities). That is the reason of trying to just the bug and cross fingers that will be accepted. 

Thanks for your support and patiente.

elupus commented:
Thanks!

javicalle commented:
Thanks Joakim
